File: algorithms/0135.candy/candy.cpp
#include <numeric>
class Solution {
public:
    int candy(vector<int>& ratings) {
        if (ratings.size() < 2) {
            return ratings.size();
        }
        vector<int> number(ratings.size(), 1);
        for (int i = 1; i < ratings.size(); ++i) {
            if (ratings[i] > ratings[i-1]) {
                number[i] =  number[i-1] + 1;
            }
        }
        for (int i = ratings.size() - 1; i > 0; --i) {
            if (ratings[i] < ratings[i-1]) {
                number[i-1] = max(number[i-1], number[i] + 1);
            }
        }
        return accumulate(number.begin(), number.end(), 0);
    }
};

File: algorithms/0135.candy/candy.rs
use std::cmp::max;

impl Solution {
    pub fn candy(ratings: Vec<i32>) -> i32 {
        let mut candies = vec![1i32; ratings.len()];
        for i in 1..ratings.len() {
            if ratings[i-1] < ratings[i] {
                candies[i] = max(candies[i-1] + 1, candies[i]);
            }
        }
        
        for i in (0..ratings.len()-1).rev() {
            if ratings[i] > ratings[i+1] {
                candies[i] = max(candies[i+1] + 1, candies[i]);
            }
        }
        candies.iter().fold(0, |x, n| x+n)
    }
}